File: LeetCode/Kai/94BinaryTreeInorderTraversal.java
//Recursive
public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new LinkedList<Integer>();
        inOrder(root, result);
        return result;
    }
    
    public void inOrder(TreeNode root, List<Integer> result) {
        if(root!=null && root.left!=null) inOrder(root.left, result);
        if(root!=null) result.add(root.val);
        if(root!=null && root.right!=null) inOrder(root.right, result);
    }
}



//Iterative
public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new LinkedList<Integer>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode node = root;
        
        while(node!=null || !stack.isEmpty()) {
            while(node!=null) {
                stack.push(node);
                node = node.left;
            }
            node = stack.pop();
            result.add(node.val);
            node = node.right;
        }
        return result;
    }
}

File: LeetCode/Liang/94BinaryTreeInorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        stack<TreeNode*> nodeStk;
        vector<int> result;
        while (root || !nodeStk.empty()) {
            // Put the left most branch into stack
            while (root) {
                nodeStk.push(root);
                root = root->left;
            }
            // Traverse the left tree;
            TreeNode *top = nodeStk.top();
            nodeStk.pop();
            result.push_back(top->val);
            root = top->right;
        }
        return result;
    }
};
